h2/src/main/org/h2/command/dml/Query.java

```diff
@@ -348,7 +350,7 @@ public final ResultInterface query(int limit, ResultTarget target) {
         }
         fireBeforeSelectTriggers();
         if (noCache || !session.getDatabase().getOptimizeReuseResults() ||
-                session.isLazyQueryExecution()) {
+                lazy) {
```

Comment:
I believe for this exact case `neverLazy` for subquery must be true, so it would be enough to check it in addition to `session.isLazyQueryExecution()` to allow result caching.

Reply:
So this would be the correct change in  h2/src/main/org/h2/command/dml/Query.java?

        if (noCache || !session.getDatabase().getOptimizeReuseResults() ||
                (session.isLazyQueryExecution() && !neverLazy)) {

Reply:
What would be the best way to add performance tests for lazy mode execution.

Reply:
Yes, it should work. 

To test it you can try to use a subquery with a deterministic function:
```sql 
select 1 from SYSTEM_RANGE(1, 10) where x in (select a from MY_DETERMINISTIC_FUNC())
```
where `MY_DETERMINISTIC_FUNC` throws an error if it is called more than once.

Reply:
Where should I put this test? h2/src/test/org/h2/test/testScript.sql

Reply:
May I ask for the help with the test. The first draft of test below does not show any difference between fixed and original code.  I could see that the condition is executed differently, but the output is the same in both cases.

```
    @Test
    public void subQueryCache() throws Exception {
        Statement stat = conn.createStatement();
        stat.execute("create alias \"ONCE\" DETERMINISTIC for \"" + this.getClass().getName() + ".once()\"");
        String s1 = "select 123 from SYSTEM_RANGE(1, 10) where x in (select id from once())";
        try (ResultSet rs = stat.executeQuery(s1)) {
            int n =0;
            while(rs.next()){
                n++;
            }
            System.out.println("n = " + n);
        }
    }

    static int count = 0;

    @SuppressWarnings("unused")
    public static SimpleResultSet once() {
        // called twice within parser and then once during execution
        if (count > 3)
            throw new IllegalArgumentException("called twice, query is not cached");
        count++;
        System.out.println("count = " + count);
        SimpleResultSet rs = new SimpleResultSet();
        rs.addColumn("ID", Types.INTEGER, 10, 0);
        for (int i = 1; i <= 10; i++)
            rs.addRow(i);
        return rs;
    }

```

Reply:
You mean the test never fail? May be you just forgot to enable lazy query execution?

Otherwise I would recommend to use debugger and look what happens.

Reply:
It looks like H2 is smart enough to optimize away call to a deterministic function and replace it with value. Then it makes sense to use custom `TableEngine` instead of a function here.